docs/client-conventions.md

```diff
@@ -0,0 +1,164 @@
+to `True`; this indicates the operation was a success. If reflecting the spec in
+the serving state is impossible, the `Ready` condition will flip to `False`;
+this indicates the operation was a failure, and the message of the status
+condition should indicate something in English about why (and the Reason field
+can indicate an emumeration suitable for i18n). Either `True` or `False`
+indicates the operation is complete, for better or worse.
+
+Note that someone else could start another operation while the cient was waiting
+for its operation. A conventional client still waits for the `Ready` condition
+to land at `True` or `False`, and then describes to the user what happened using
+logic based on the intended effect.
+
+For example:
+ * Client A deploys image `gcr.io/foods/vegetables:eggplant`
+ * While that is not yet Ready, client B deploys `gcr.io/foods/vegetables:squash`
+ * The `eggplant` revision becomes Ready: True, and the service moves traffic to it.
+ * The `squash` revision fails to bind to a port, and becomes Ready: False
+ * The Service switches from Ready: Unknown to Ready: False because `squash` failed.
+
+Both client A and B should wait for the last step in this procedure.
+ * Client A sees that latestReadyRevision is the revision with the
+   [nonce](#associate-modifications-with-revisions) it specified, and that
+   `latestCreatedRevision` is not. It tells the user that deploying was
+   successful.
+ * Client B sees that `latestCreatedRevision` is the revision with the nonce it
+   specified; it reports the failure with the appropriate message.
+
+The rule is "Wait for `Ready` to become `True` or `False`, then report on
+whether your intent was accomplished". The `Ready` success or failure can be
+part of this report, but may be confusing (as in the example) if it's the only
+thing you report.
+
+## Associate modifications with Revisions
+
+Every time the client changes a Service or Configuration in a way that results
+in a new Revision, it should change the label `client.knative.dev/nonce` to a
+new random value, chosen from a wide enough space to avoid collisions.
+
+This way, the client can make a label selector query for Revisions with this
+nonce value to find the Revision the particular change generated. The client can
+use that revision to, for example, inform the user about the readiness of their
+requested change.
+
+### In the presence of non-conventional clients
+
+If an unconventional client neglects to reset this value, the label selector
+query could result in more than one revision; the client should pick the one
+with the lowest `configurationGeneration` in this case.
+
+## Force creation of a new Revision
+
+The way to deploy new code with a previously-used tag is to make a new Revision,
+which the Revision controller will re-pull and lock it to the current image at
+that tag. Unfortunately Knative won't create a new Revision if you don't change
+the Configuration.
```

Comment:
I would rephrase this last sentence as: Declarative APIs require some change to the desired state of the world (e.g. `spec`) to enact change, or the controllers have nothing to observe.